pkg/rules/manager_test.go

```diff
@@ -397,3 +399,57 @@ groups:
 	testutil.Ok(t, err)
 	testutil.Equals(t, 0, len(thanosRuleMgr.RuleGroups()))
 }
+
+func TestManagerRunRulesWithRuleGroupLimit(t *testing.T) {
+	dir, err := ioutil.TempDir("", "test_rule_rule_groups")
+	testutil.Ok(t, err)
+	defer func() { testutil.Ok(t, os.RemoveAll(dir)) }()
+	filename := filepath.Join(dir, "with_limit.yaml")
+	testutil.Ok(t, ioutil.WriteFile(filename, []byte(`
+groups:
+- name: "something1"
+  interval: 1ms
+  limit: 1
+  rules:
+  - alert: "some"
+    expr: "up>0"
+    for: 0s
+`), os.ModePerm))
+
+	thanosRuleMgr := NewManager(
+		context.Background(),
+		nil,
+		dir,
+		rules.ManagerOptions{
+			Logger:    log.NewLogfmtLogger(os.Stderr),
+			Queryable: nopQueryable{},
+		},
+		func(partialResponseStrategy storepb.PartialResponseStrategy) rules.QueryFunc {
+			return func(ctx context.Context, q string, ts time.Time) (promql.Vector, error) {
+				return []promql.Sample{
+					{
+						Point:  promql.Point{T: 0, V: 1},
+						Metric: labels.FromStrings("foo", "bar"),
+					},
+					{
+						Point:  promql.Point{T: 0, V: 1},
+						Metric: labels.FromStrings("foo1", "bar1"),
+					},
+				}, nil
+			}
+		},
+		nil,
+		"http://localhost",
+	)
+	thanosRuleMgr.Run()
+	defer func() {
+		thanosRuleMgr.Stop()
+	}()
+	testutil.Ok(t, thanosRuleMgr.Update(time.Millisecond, []string{filename}))
+	testutil.Equals(t, 1, len(thanosRuleMgr.protoRuleGroups()))
+	testutil.Equals(t, 1, len(thanosRuleMgr.protoRuleGroups()[0].Rules))
+	testutil.Equals(t, string(rules.HealthUnknown), thanosRuleMgr.protoRuleGroups()[0].Rules[0].GetAlert().Health)
+	time.Sleep(time.Millisecond * 5)
```

Comment:
Please don't use `time.Sleep` in tests. It will lead to flakiness in CI. I suggest on using `WaitSumMetrics` to wait for at least one evaluation to happen (:

Reply:
Nice idea.

Reply:
Done